packages/contents/src/drivecontent.ts

```diff
@@ -0,0 +1,235 @@
+import { PathExt } from '@jupyterlab/coreutils';
+import { Contents } from '@jupyterlab/services';
+import {
+  BLOCK_SIZE,
+  IDriveRequest,
+  IDriveRenameRequest,
+  IDriveMknodRequest,
+  IDrivePutRequest,
+  DriveGetmodeResponse,
+  DriveLookupResponse,
+  DriveReaddirResponse,
+  DriveGetattrResponse,
+  DriveGetResponse,
+} from './drivefs';
+import { DIR_MODE, FILE_MODE } from './emscripten';
+
+export interface IDriveContentsProcessor {
+  /**
+   * Process a content request
+   *
+   * @param request the request
+   */
+  processDriveRequest(request: IDriveRequest): Promise<void>;
+
+  /**
+   * Process the request to read a directory content
+   *
+   * @param request the request
+   */
+  readdir(request: IDriveRequest): Promise<DriveReaddirResponse>;
+
+  /**
+   * Process the request to remove a directory
+   *
+   * @param request the request
+   */
+  rmdir(request: IDriveRequest): Promise<void>;
+
+  /**
+   * Process the request to rename a file or directory
+   *
+   * @param request the request
+   */
+  rename(request: IDriveRenameRequest): Promise<void>;
+
+  /**
+   * Process the request to get the node mode (file or directory)
+   *
+   * @param request the request
+   */
+  getmode(request: IDriveRequest): Promise<DriveGetmodeResponse>;
+
+  /**
+   * Process the request to check if a node exist
+   *
+   * @param request the request
+   */
+  lookup(request: IDriveRequest): Promise<DriveLookupResponse>;
+
+  /**
+   * Process the request to create a directory/file
+   *
+   * @param request the request
+   */
+  mknod(request: IDriveMknodRequest): Promise<void>;
+
+  /**
+   * Process the request to get a node stats
+   *
+   * @param request the request
+   */
+  getattr(request: IDriveRequest): Promise<DriveGetattrResponse>;
+
+  /**
+   * Process the request to get the content of a file
+   *
+   * @param request the request
+   */
+  get(request: IDriveRequest): Promise<DriveGetResponse>;
+
+  /**
+   * Process the request to write the content of a file
+   *
+   * @param request the request
+   */
+  put(request: IDrivePutRequest): Promise<void>;
+}
+
+/**
+ * Class for processing a drive request from the DriveFS.
+ */
+export class DriveContentsProcessor implements IDriveContentsProcessor {
+  private contentsManager: Contents.IManager;
+
+  constructor(options: DriveContentsProcessor.IOptions) {
+    this.contentsManager = options.contentsManager;
+  }
+
+  async processDriveRequest(request: IDriveRequest): Promise<any> {
```

Comment:
Sooo... there's almost certainly a way to use a type discriminator to get well-typed values out, removing the need for `!` in a few places elsewhere.

Basically:

```ts
// somewhere, maybe in `tokens`?
export type TDriveMethod = 'readdir' | /* ... */ | 'put';
// somewhere, else maybe _also_ in `tokens`?
export interface IDriveRequest {
  method: TDriveMethod;
  // ...
}
```

Then each of the requests would refine it down to _exactly_ a member;

It's also possible to make wacky types that work as lookup tables, but I can't pull it out off the top of my head..

Reply:
Sorry, I don't understand your comment, can you elaborate?

The `IDriveRequest` is already defined that way on the main branch: https://github.com/jupyterlite/jupyterlite/blob/main/packages/contents/src/drivefs.ts#L31

Reply:
Ok, cool, we have that. My point is we should not need `!` in new places elsewhere, as the return types should be inferrable from the inputs, based on `method`... seeing if i can gin up something more concrete...

Reply:
Ah ok I see what you mean, I'll have a look

Reply:
Ooof... I _have_ done something like this:

https://github.com/jupyter-lsp/jupyterlab-lsp/pull/278/files#diff-7c7804a4b99d4b67b07ab518d033a6de85bdbf291e5e03400106901ff2501d5aR30

(most of the types in that signature are in that PR, but it's still a beast)

We'd basically want types for each of the methods (that actually overload the `method` with the actual value), then would be able to reference them with `<T, U, V>`. It's a lot, but adding _new_ APIs with `any` feels bad.

Reply:
Thanks for the reference! I just pushed an attempt to implement that. It seems to compile?